store/sqlstore/channel_store.go

```diff
@@ -2254,6 +2254,7 @@ func (s SqlChannelStore) MigrateChannelMembers(fromChannelId string, fromUserId
 
 		if len(channelMembers) == 0 {
 			// No more channel members in query result means that the migration has finished.
+			transaction.Rollback()
```

Comment:
There's a couple of other returns in this method where we might leave the transaction open if an occur should occur. I'm wondering if we should add a 
```go
defer transaction.Rollback()
```

on line 2248 after beginning the transaction successfully, and ensuring we close the transaction in all cases. As per the implementation in gorp, the Rollback won't do anything if the transaction was successfully committed (other than return an error, which we can ignore):
https://github.com/go-gorp/gorp/blob/f3677d4a0a8838c846ed41bf41927f2c8713bd60/transaction.go#L163-L174

If this makes sense, we should do it in the other store as well... and I'm wondering if we ought to sweep the code base and do it everywhere? Happy to accept this PR with the `defer` just for the methods in question, but if this approach makes sense and you're up for task, feel free to fix this everywhere.

Reply:
Hi! Thanks for the feedback @lieut-data!

As per your recommendation, I've just committed another patch to use `defer transaction.Rollback()` instead. I went through all files in `store/sqlstore` and looked for calls to the `gorp.DbMap` `Begin()` method. I found 25 of them, which I all patched accordingly. Most of the times, the `defer` was unnecessary as all conditional branches where treated correctly (e.g., see [`SqlSupplier.SchemeSave()`](https://github.com/mattermost/mattermost-server/blob/f42c00ee5331e4e97b57b09b81ea73728b3566ab/store/sqlstore/scheme_supplier.go#L34-L67) in `scheme_supplier.go`), but I still added the `defer` part, as an additional safety precaution. As you wrote, calling it twice won't hurt anyway.

I also realized that the `defer` strategy was not enough in the [`UpgradeDatabaseToVersion33()`](https://github.com/mattermost/mattermost-server/blob/f42c00ee5331e4e97b57b09b81ea73728b3566ab/store/sqlstore/upgrade.go#L159-L234) method of `upgrade.go`: On error, this method calls [`themeMigrationFailed()`](https://github.com/mattermost/mattermost-server/blob/f42c00ee5331e4e97b57b09b81ea73728b3566ab/store/sqlstore/upgrade.go#L153-L157), which in turn calls `os.Exit()`, bypassing all `defer`'ed calls to `transaction.Rollback()`. Therefore, I've added them explicitly before each call to `themeMigrationFailed()`.

Hopefully, this should be enough to prevent stale opened transactions.